LPDeviceManager.device.lib.nut

```diff
@@ -0,0 +1,340 @@
+     */
+    constructor(cm, handlers = {}, isDebug = false) {
+        const __RM_CALLBACK_NAME = "lp-callback";
+
+        _cm = cm;
+        _onIdleCbs = [];
+        _wakeupReason = hardware.wakereason();
+        _isDebug = isDebug;
+
+        _onTimer = ("onTimer" in handlers) ? handlers.onTimer : null;
+        _onSwReset = ("onSwReset" in handlers) ? handlers.onSwReset : null;
+        _onColdBoot = ("onColdBoot" in handlers) ? handlers.onColdBoot : null;
+        _onInterrupt = ("onInterrupt" in handlers) ? handlers.onInterrupt : null;
+        _defaultOnWake = ("defaultOnWake" in handlers) ? handlers.defaultOnWake : null;
+        _onPowerRestored = ("onPowerRestored" in handlers) ? handlers.onPowerRestored : null;
+
+        imp.wakeup(0, _dispatchEvents.bindenv(this));
+    }
+
+    /**
+     * Adds an action to be called on idle. The imp should not go to sleep until there is a least one
+     * item in the onIdle list.
+     */
+    function addOnIdle(callback) {
+        // reset the onidle handler
+        imp.onidle(_processOnIdle.bindenv(this));
+
+        if (_onIdleCbs.find(callback) != null) {
+            return; // the item already exists
+        }
+        if (_isFunc(callback)) {
+            _onIdleCbs.append(callback);
+            return;
+        }
+        _err("Invalid callback type");
+    }
+
+    /**
+     * Synchronously executes the specified action(s) and goes to deep sleep for the specified period of time.
+     *
+     * @param {function} actions - an action function to be fulfilled before we go to sleep. The function takes no arguments.
+     * @param {integer} sleepTime - specified the time the device is going to sleep for after the actions are all fulfiled
+     */
+    // TODO: needs a lot of good examples covering all the edge use cases
+    function doAndSleep(action, sleepTime) {
+        _isFunc(action) && action();
+        sleepFor(sleepTime);
+    }
+
+    /**
+     * Schedules asynchronous execution of the specified action(s),
+     * when the actions are done, goes to deep sleep for the specified period of time.
+     * The device goes to sleep on whichever earlier occurs: the timeout or the action has completed.
+     *
+     * @param {action-callback} action - an action function to be fulfilled before we go to sleep
+     * @param {float} sleepTime - specified the time the device is going to sleep for after the actions are all fulfiled
+     * @param {float} timeout - the action timeout
+     */
+    /**
+     * The function that implements an action to be performed.
+     *
+     * @callback action-callback
+     * @param {function} done - the function to be called by the action when done.
+     */
+    function doAsyncAndSleep(action, sleepTime, timeout = LP_DEFAULT_TIMEOUT_SEC) {
+        // Schedule deep sleep after timeout
+        imp.wakeup(timeout, function() {
+            sleepFor(sleepTime);
+        });
```

Comment:
missing bindenv

Reply:
Thank you @betzrhodes, good catch. Fixed.